[feature][riscv] handle target address calculation in llvm-objdump disassembly for riscv

arjunUpatel opened the pull request:
Resolves #108469

arjunUpatel committed:
added evaluateInstruction method. needs tests

arjunUpatel committed:
dealing with git

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -230,6 +236,44 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+      case RISCV::ADDI:
+      case RISCV::ADDIW: {
+        if (auto TargetRegState = getGPRState(Inst.getOperand(1).getReg())) {
+          Target  = *TargetRegState + Inst.getOperand(2).getImm();
```
ADDIW sign extends the result.

arjunUpatel replied:
The sign extension is handled internally since the the immediate value has type `int64_t` which according to https://en.cppreference.com/w/cpp/types/integer is guaranteed to use 2s complement with no padding, which is essentially what sign extension does. I went ahead and printed the bit representation of the immediate values passed to the instructions and they were sign extended as per specification.

topperc replied:
ADDIW needs to sign extend the addition result.

arjunUpatel replied:
I am still a bit confused by this feedback, mainly because to my understanding, the sign extension of the result is being handled by the fact that the values are loaded in memory already sign extended to 64 bits along with 2s complement addition preserving the sign extension over arithmetic manipulation.

Here is my understanding so far:
1. A signed value, extended to 64 bits, is loaded into `*TargetRegState` by either AUIPC or LUI
2. The immediate value, also already sign extended to 64 bits, is added to `*TargetRegState`
3. 2s complement addition is carried out. The sum will stay sign extended to 64 bits as per 2s complement

Is this correct or am I missing something?

topperc replied:
ADDIW adds bit 31:0 of both inputs, calculates a 32 bit result then sign extends that to 64 bits. Bits 63:32 of the inputs are discarded and don't affect the result. This is different than adding 2 int64_t values together.

arjunUpatel committed:
sign extend relevant immediates

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -178,21 +180,35 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+        MCRegister Reg = Inst.getOperand(0).getReg();
+        if (Reg == RISCV::X2)
+          break;
+        setGPRState(Reg, SignExtend64<17>(Inst.getOperand(1).getImm() << 12));
```
This should be 18 not 17. The immediate is 6 bits placed in bits 17:12.

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -244,34 +254,43 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+          uint64_t mask = ~(0) >> XLEN;
+          Target  = *TargetRegState + SignExtend64<12>(Inst.getOperand(2).getImm());
+          Target &= mask;
+          Target = SignExtend64<32>(Target);
```
This SignExtend shouldn't happen for ADDI. Only ADDIW. They are different instructions and have different behavior.

arjunUpatel committed:
fix indentation

arjunUpatel committed:
fix ADDI + sign extention bugs

arjunUpatel committed:
prevent symbol reoslution in empty sections

arjunUpatel committed:
added support for compressed instructions

arjunUpatel committed:
Merge branch 'llvm:main' into main

arjunUpatel committed:
call evaluateInstruction only when target=RISCV

arjunUpatel committed:
fix bug cause test failures with build

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -178,21 +180,35 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+        MCRegister Reg = Inst.getOperand(0).getReg();
+        if (Reg == RISCV::X2)
+          break;
+        setGPRState(Reg, SignExtend64<18>(Inst.getOperand(1).getImm() << 12));
```
I think c.lui immediate is always in the range [0xfffe0, 0xfffff] or [1, 31] so SignExtend<32> like RISCV::LUI should work.

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -178,21 +180,35 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
-        auto DefReg = Inst.getOperand(I).getReg();
-        if (isGPR(DefReg))
-          setGPRState(DefReg, std::nullopt);
+      case RISCV::LUI: {
```
`case` should have the same indent level as the proceeding `switch` per coding standards.

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -230,6 +246,97 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+        if (auto TargetRegState = getGPRState(Inst.getOperand(1).getReg())) {
+          uint64_t Mask = ~((uint64_t)0) >> 32;
+          Target  = *TargetRegState + SignExtend64<12>(Inst.getOperand(2).getImm());
+          Target &= Mask;
```
Why is this `&=` line needed?

topperc replied:
But the SignExtend64<32>(TargeT) right after the &= overwrites the bits anyway

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -230,6 +246,97 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+        if (auto TargetRegState = getGPRState(Inst.getOperand(1).getReg())) {
+          // TODO: Figure out ways to find the actual value of XLEN during analysis
+          int XLEN = 32;
+          uint64_t Mask = ~((uint64_t)0) >> (64 - XLEN);
```
Use `maskTrailingOnes` from MathExtras.h

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -230,6 +246,97 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+          // TODO: Figure out ways to find the actual value of XLEN during analysis
+          int XLEN = 32;
+          uint64_t Mask = ~((uint64_t)0) >> (64 - XLEN);
+          Target = *TargetRegState + SignExtend64<12>(Inst.getOperand(2).getImm());
```
Is the `SignExtend64<12>` needed? Isn't the constant stored in sign extended form?

arjunUpatel replied:
I think so. I was going off of suggestions provided in a previous [review](https://github.com/llvm/llvm-project/pull/109914#discussion_r1775677211) where we concluded that the immediates of LUI and AUIPC should be sign extended and extrapolated that decision to the rest of the instructions. Any holes in my understanding as to why those should be sign extended but not the immediates of ADDI in this case (and other instructions)?

I ended up digging through the RISCV spec and noticed it specifically states that 
> ADDI adds the sign-extended 12-bit immediate ...

this was true for other instructions as well, but not AUIPC and LUI. Following this distinction, I am assuming that the immediate values for AUIPC and LUI are not stored in sign extended form (while others are) leading to the sign extension being needed for these instructions and not others. Any insights appreciated :)

topperc replied:
That's correct. LUI and AUIPC immediates are 20-bit unsigned values. The 12-bit immediates for ADDI, XORI, ANDI, ORI, loads, stores are always in sign extended form.

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -230,6 +246,97 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+      case RISCV::ADDIW: {
+        if (auto TargetRegState = getGPRState(Inst.getOperand(1).getReg())) {
+          uint64_t Mask = ~((uint64_t)0) >> 32;
+          Target  = *TargetRegState + SignExtend64<12>(Inst.getOperand(2).getImm());
```
Is the `SignExtend64<12>` needed? Isn't the constant stored in sign extended form?

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -230,6 +246,97 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+        if (auto TargetRegState = getGPRState(Inst.getOperand(1).getReg())) {
+          uint64_t Mask = ~((uint64_t)0) >> 32;
+          Target &= Mask;
+          Target = *TargetRegState + SignExtend64<6>(Offset);
```
Is the `SignExtend64<6>` needed? Isn't the constant stored in sign extended form?

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -230,6 +246,97 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+      case RISCV::FSH:
+      case RISCV::FSW:
+      case RISCV::FSD: {
+        int64_t Offset = SignExtend64<12>(Inst.getOperand(2).getImm());
```
Is the `SignExtend64<12>` needed? Isn't the constant stored in sign extended form?

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -230,6 +246,97 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+        if (Offset == 0)
+          break;
+        if (auto TargetRegState = getGPRState(Inst.getOperand(1).getReg())) {
+          uint64_t Mask = ~((uint64_t)0) >> 32;
```
Use maskTrailingOnes

topperc commented on llvm/tools/llvm-objdump/llvm-objdump.cpp:
```diff
@@ -2479,7 +2480,7 @@ disassembleObject(ObjectFile &Obj, const ObjectFile &DbgObj,
               } else if (LabelAvailable) {
                 *TargetOS << " <" << AllLabels[Target] << ">";
-              }
+              } 
```
Extra space at end of line?

arjunUpatel committed:
Instr eval based on reg width and attempt to pass tests

lenary commented on llvm/include/llvm/MC/MCInstrAnalysis.h:
```diff
@@ -181,6 +181,12 @@ class MCInstrAnalysis {
+  /// the target address. Return true success, and the address in Target.
+  virtual bool
+  evaluateInstruction(const MCInst &Inst, uint64_t Addr, uint64_t Size,
+                      uint64_t &Target, int ArchRegWidth) const;
```
The `ArchRegWidth` parameter here feels strange given it shouldn't really change between instructions. I realise changing the signature of `Target::MCInstrAnalysisCtorFnTy` to pass the triple (most other construction functions have some kind of access to it) is pretty invasive and so out of the question - maybe instead we could pass the width to the constructor and use two different `createRISCVInstrAnalysis` functions, one for rv32 and one for rv64? Then we wouldn't need it as a parameter.

arjunUpatel replied:
Makes sense. On it :saluting_face:

arjunUpatel committed:
address comments, merge instruction evaluation and pass register width through constructor

arjunUpatel committed:
remove debugging code

arjunUpatel committed:
silly me forgot to save changes

arjunUpatel committed:
run clang format

lenary commented:
This implementation looks good, but it really needs tests :)

arjunUpatel committed:
fix code suggestion

arjunUpatel committed:
remove absolute first

arjunUpatel commented:
I am consistently failing one of the tests and have rattled my brain quite a bit about it. Here is whats going + guidance needed:

The test I am failing is the 5th test in llvm-project/llvm/test/tools/llvm-objdump/X86/disassemble-same-section-addr.test. This particular test ensures that the absolute symbol is used for target address resolution if no symbol is found in the candidate section. Here is the structure of the test ELF file:

Sections:
  - Name:    .caller
    Type:    SHT_PROGBITS
    Flags:   [SHF_ALLOC, SHF_EXECINSTR]
    Address: 0x0
    Content: e800000000 ## Call instruction to next address.
  - Name:    .first
    Type:    SHT_PROGBITS
    Flags:   [SHF_ALLOC, SHF_EXECINSTR]
    Address: 0x5
    Size:    [[SIZE1]]
  - Name:    .second
    Type:    SHT_PROGBITS
    Flags:   [SHF_ALLOC, SHF_EXECINSTR]
    Address: 0x5
    Size:    [[SIZE2]]

Symbols:
  - Name:    target
    Section: [[SECTION]]
    Value:   0x5
  - Name:    other
    Index:   [[INDEX]]
    Value:   0x0

Here is the disassembly of the file when the parameters are set in the following way (according to test 5):
SIZE1=1, SIZE2=0, SECTION=.caller, INDEX=SHN_ABS

Disassembly of section .caller:
0000000000000000 <.caller>:
       0: e8 00 00 00 00               	callq	0x5

Disassembly of section .first:
0000000000000005 <.first>:
       5: 00                           	<unknown>

Target address resolution shall occur at address the callq instruction in the .caller section. Currently, the expected resolution is <other+0x5>. This would occur because the current implementation strictly checks the set of sections with the same address where the address is the closest to and less than or equal to the target. Lets call this set of sections $A$. In the current implementation if all the sections in $A$ are empty, then the absolute symbol is used. But what about symbols that occur in sections before $A$? Should the sections before $A$ also not be checked for valid symbols and the address resolution printed relative to one of those symbols? Following this scheme, the address resolution in this case will switch to <target> since the address of target is 0x5.

Checking sections before $A$ is exactly what binutils seems to be doing leading to discrepancies and failing tests when trying to fully mimic its behavior. Which scheme should I proceed with?

arjunUpatel committed:
objdump prioritize actual symbols over dummy symbols during resolution

arjunUpatel committed:
remove unist.h from includes

arjunUpatel committed:
Revert "objdump prioritize actual symbols over dummy symbols during resolution"

This reverts commit b78131200f77913fa9cd4c982e39d24274e8ccee.

arjunUpatel committed:
modify test to effectively test new functionailty

arjunUpatel committed:
Update .gitignore

arjunUpatel committed:
Update tests to match new functionality

PiJoules commented:
👋 How's progress on this? Is there anything else needed for reviews?

topperc commented on .gitignore:
```diff
@@ -28,6 +28,7 @@
 
 # Nested build directory
 /build*
+/local-tests
```
I don't think we should be changing this file in this patch.

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -161,8 +164,8 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
 public:
-  explicit RISCVMCInstrAnalysis(const MCInstrInfo *Info)
-      : MCInstrAnalysis(Info) {}
+  explicit RISCVMCInstrAnalysis(const MCInstrInfo *Info, int ArchRegWidth)
```
Use `unsigned` instead of `int` if its never negative.

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -230,6 +240,64 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
+    case RISCV::C_FLD:
+    case RISCV::C_FSD:
+    case RISCV::C_SW:
+    case RISCV::C_LW:
```
Missing compressed loads/stores from Zcb extension. C_LBU, C_LH, C_LHU, C_SB, C_SH.

Missing C_LWSP, C_SWSP, C_LDSP, and C_SDSP from C extension. These implicitly use SP as the base register.

arjunUpatel committed:
Add tests of scenarios provided in issue description (see issue related to pr)

arjunUpatel committed:
Update tests to accurately match new symbol resolution search pattern

arjunUpatel committed:
Added tests to increase code coverage of new functionality

arjunUpatel committed:
Help llvm-lit find new tests

arjunUpatel committed:
FIx zero register bug. Previously address resolution would be triggered if the input register was the zero register.

arjunUpatel committed:
Remove ignore of local folder as per comments

arjunUpatel committed:
Remove extraneous header as per comments

arjunUpatel committed:
Use unsigned instead of signed int for values that are always positive - per comments

arjunUpatel committed:
Add support for Zcb extensions + corresponding tests

arjunUpatel committed:
Added support for stack pointer based load and stores

arjunUpatel committed:
Merge branch 'main' into main

arjunUpatel committed:
Use unsigned int for ArchRegWidth

The variable is never negative. The change here simply matches the type of the variable passed into the constructor.

arjunUpatel committed:
Update tests to match new functionality

The way sections are searched for symbols to print for address reolution was  changed in this patch causing these tests to fail. This commit updates the expected output of the tests to match the new functionality.

arjunUpatel committed:
One more try at passing tests

arjunUpatel committed:
Non-exact offset match for failing test

While trying to pass tests, multiple different values were emitted for the offset changed in this commit. This change allows the offset to be non-exact so that the test case does not break unexpectedly in future builds. Between https://buildkite.com/llvm-project/github-pull-requests/builds/181865#0196f537-b4c8-4595-9e8c-9a240d9a07a5 and https://buildkite.com/llvm-project/github-pull-requests/builds/181752#0196f409-1e95-4eea-9f3f-500c0099983c/6-10, in the former the the expected offset was  fc58 while in the letter the expected offset was fc68. Locally the objdump output yielded fc78.

topperc commented on llvm/include/llvm/MC/MCInstrAnalysis.h:
```diff
@@ -181,6 +181,11 @@ class MCInstrAnalysis {
   evaluateBranch(const MCInst &Inst, uint64_t Addr, uint64_t Size,
                  uint64_t &Target) const;
 
+  /// Given an instruction that accesses a menory address, try to compute
```
menory -> memory

topperc commented on llvm/include/llvm/MC/MCInstrAnalysis.h:
```diff
@@ -181,6 +181,11 @@ class MCInstrAnalysis {
                  uint64_t &Target) const;
 
+  /// Given an instruction that accesses a menory address, try to compute
+  /// the target address. Return true success, and the address in Target.
```
```suggestion
  /// the target address. Return true on success, and the address in \p Target.
```

arjunUpatel committed:
Improve documentation for evaluateInstruction

arjunUpatel committed:
Merge branch 'main' into main

arjunUpatel commented:
This one is ready for final review and subsequent merge. All features and tests have been successfully implemented on my side (hopefully)

arjunUpatel committed:
Fix typo in llvm/test/tools/llvm-objdump/RISCV/riscv-ar-coverage.s documentation

Co-authored-by: Craig Topper <craig.topper@sifive.com>

arjunUpatel committed:
Avoid else case as per comments

arjunUpatel commented on llvm/tools/llvm-objdump/llvm-objdump.cpp:
```diff
@@ -1520,9 +1520,9 @@ collectLocalBranchTargets(ArrayRef<uint8_t> Bytes, MCInstrAnalysis *MIA,
-            !Targets.count(Target)) {
+        bool BranchTargetKnown = MIA->evaluateBranch(Inst, Index, Size, Target);
+        if (BranchTargetKnown && (Target >= Start && Target < End) &&
+            !Labels.count(Target)) {
```
You are right, the code should reflect `!Targets.count` instead of `!Labels.count`. It seems like in the period of my inactivity with this PR, someone else happened to converge upon the same change (making sure the container of the obtained Target is empty before adding stuff). They just hapened to implement it in a different way.

arjunUpatel committed:
Likely did a bad merge in the past. Updating to reflect correct code with same functionality

arjunUpatel committed:
Merge branch 'main' into main

arjunUpatel committed:
Merge branch 'main' of github.com:arjunUpatel/llvm-project

topperc commented on llvm/tools/llvm-objdump/llvm-objdump.cpp:
```diff
@@ -2392,14 +2393,18 @@ disassembleObject(ObjectFile &Obj, const ObjectFile &DbgObj,
-                    break;
+                  }
                   TargetSectionSymbols.push_back(&AllSymbols[It->second]);
+                  if (!AllSymbols[It->second].empty())
```
Can we avoid a second `AllSymbols[It->second]` here since we already did it for the previous line. That's a map lookup, we shouldn't do it twice.

topperc commented on llvm/tools/llvm-objdump/llvm-objdump.cpp:
```diff
@@ -2392,14 +2393,18 @@ disassembleObject(ObjectFile &Obj, const ObjectFile &DbgObj,
                       return O.first <= Target;
                     });
-                uint64_t TargetSecAddr = 0;
+                uint64_t TargetSecAddr = It == SectionAddresses.end() ? 0 : It->first;
```
Are the non-RISC-V test changes because of this change?

arjunUpatel replied:
Yes, and the following loop. Previously, the loop would only add labels in the set of sections that were closest to `Target`. _Set_ of section here because multiple sections can have the same address, so all of their symbols would be added to the set of candidate symbols. The problem here is that that sometimes this set of sections would not have any symbols, which would cause global symbols to be used for address resolution. The following changes make it such that we loop down from sections closest to the `Target` and populate the set of candidate symbols with symbols from the first set of sections that do contain symbols. This change bumps the address resolution capability to better match that of binutils

topperc replied:
This change should probably be its own PR? It's not described in the title or description. As titled, the patch looks like a riscv only change, but it's not.

arjunUpatel replied:
On me, I was aiming to match the obtained output with example expected output generated by binutils objdump in the issue description. Ill move all the non-RISCV affecting changes to another PR. Apologies for the oversight

arjunUpatel committed:
Reduce map lookup of section symbols

arjunUpatel committed:
Remove changes affecting non-RISCV targets

arjunUpatel committed:
Merge branch 'main' of github.com:arjunUpatel/llvm-project

jrtc27 commented:
Can you please give this a proper commit message? See https://llvm.org/docs/DeveloperPolicy.html#commit-messages for guidance.

arjunUpatel commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCTargetDesc.cpp:
```diff
@@ -327,8 +414,12 @@ class RISCVMCInstrAnalysis : public MCInstrAnalysis {
 
-static MCInstrAnalysis *createRISCVInstrAnalysis(const MCInstrInfo *Info) {
-  return new RISCVMCInstrAnalysis(Info);
+static MCInstrAnalysis *createRISCV32InstrAnalysis(const MCInstrInfo *Info) {
```
I think the best way to do this is to make the subtarget info a member of the parent class `MCInstrAnalysis`. The subtarget info is also used by other member functions, notably `findPltEntries`. All the uses of subtarget info seem to be static across all the member functions that need this information, that is, it does not change in subsequent calls to these functions in a given invocation. As such, it makes sense to tie it to the parent class so it is set once and referenced when needed versus continuously being passed as an argument even if it never changes. For the sake of keeping this PR relevant perhaps we can pass the subtarget info as an argument to `evaluateInstruction` for now so that it matches the implementation of `findPltEntries` and how the subtarget info is threaded in can be updated in a follow up PR?

arjunUpatel commented:
After [this discussion in this thread](https://github.com/llvm/llvm-project/pull/109914#discussion_r2136630018) a few changes had to be reverted. But because of me being a noob at git, my reversion scheme involved some not so optimal actions. So instead of force pushing and losing the context of the conversation, I am continuing the development in [this PR](https://github.com/llvm/llvm-project/pull/144620)
